File: cli/cmd/tools.go
package cmd

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/spf13/cobra"

	"openpa.local/cli/internal/output"
)

func newToolsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "tools",
		Aliases: []string{"tool"},
		Short:   "List and configure tools & skills",
	}
	cmd.AddCommand(
		newToolsListCmd(),
		newToolsGetCmd(),
		newToolsEnableCmd(true),
		newToolsEnableCmd(false),
		newToolsSetVarCmd(),
		newToolsSetArgsCmd(),
		newToolsSetLLMCmd(),
		newToolsResetLLMCmd(),
	)
	return cmd
}

func newToolsListCmd() *cobra.Command {
	var typeFilter string
	cmd := &cobra.Command{
		Use:   "list",
		Short: "List all tools (built-in, mcp, a2a, skill, intrinsic)",
		RunE: func(cmd *cobra.Command, _ []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			tools, err := state.Client.ListTools(cmd.Context(), typeFilter)
			if err != nil {
				return err
			}
			if state.Output.JSON {
				return output.PrintJSON(tools)
			}
			t := output.NewTable(state.Output, "TOOL_ID", "TYPE", "ENABLED", "CONFIGURED", "NAME")
			for _, row := range tools {
				t.AddRow(
					stringField(row, "tool_id"),
					stringField(row, "tool_type"),
					boolField(row, "enabled", true),
					boolField(row, "configured", false),
					stringField(row, "name"),
				)
			}
			t.Render()
			return nil
		},
	}
	cmd.Flags().StringVar(&typeFilter, "type", "",
		"Filter by tool_type: built-in, mcp, a2a, skill, intrinsic")
	return cmd
}

func newToolsGetCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "get <tool_id>",
		Short: "Show detailed configuration for a tool",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			tool, err := state.Client.GetTool(cmd.Context(), args[0])
			if err != nil {
				return err
			}
			if state.Output.JSON {
				return output.PrintJSON(tool)
			}
			renderToolDetail(tool)
			return nil
		},
	}
}

func renderToolDetail(t map[string]any) {
	output.PrintKV([][2]string{
		{"tool_id", stringField(t, "tool_id")},
		{"name", stringField(t, "name")},
		{"tool_type", stringField(t, "tool_type")},
		{"description", stringField(t, "description")},
		{"configured", boolField(t, "configured", false)},
	})
	if cfg, ok := t["config"].(map[string]any); ok {
		fmt.Println()
		fmt.Println("--- config ---")
		dumpJSON(cfg)
	}
	if locked, ok := t["locked_llm_fields"].([]any); ok && len(locked) > 0 {
		fmt.Println()
		fmt.Print("locked_llm_fields: ")
		parts := make([]string, 0, len(locked))
		for _, v := range locked {
			parts = append(parts, fmt.Sprintf("%v", v))
		}
		fmt.Println(strings.Join(parts, ", "))
	}
}

func newToolsEnableCmd(enable bool) *cobra.Command {
	use := "enable"
	short := "Enable an A2A or MCP tool"
	if !enable {
		use = "disable"
		short = "Disable an A2A or MCP tool"
	}
	return &cobra.Command{
		Use:   use + " <tool_id>",
		Short: short,
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.SetToolEnabled(cmd.Context(), args[0], enable)
		},
	}
}

func newToolsSetVarCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "set-var <tool_id> KEY=VALUE [KEY=VALUE...]",
		Short: "Set Tool Variables (env-style key/value pairs)",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			vars := map[string]string{}
			for _, kv := range args[1:] {
				k, v, ok := strings.Cut(kv, "=")
				if !ok {
					return fmt.Errorf("expected KEY=VALUE, got %q", kv)
				}
				vars[k] = v
			}
			return state.Client.SetToolVariables(cmd.Context(), args[0], vars)
		},
	}
}

func newToolsSetArgsCmd() *cobra.Command {
	var jsonBody string
	cmd := &cobra.Command{
		Use:   "set-args <tool_id> --json '<JSON object>'",
		Short: "Set Tool Arguments from a JSON object",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			if jsonBody == "" {
				return fmt.Errorf("--json is required")
			}
			var values map[string]any
			if err := json.Unmarshal([]byte(jsonBody), &values); err != nil {
				return fmt.Errorf("invalid --json: %w", err)
			}
			return state.Client.SetToolArguments(cmd.Context(), args[0], values)
		},
	}
	cmd.Flags().StringVar(&jsonBody, "json", "", "Tool arguments as a JSON object")
	return cmd
}

func newToolsSetLLMCmd() *cobra.Command {
	var (
		provider, model, reasoning string
		fullReasoning              string
	)
	cmd := &cobra.Command{
		Use:   "set-llm <tool_id>",
		Short: "Set LLM Parameters for a tool (partial — omitted flags unchanged)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			params := map[string]any{}
			if provider != "" {
				params["llm_provider"] = provider
			}
			if model != "" {
				params["llm_model"] = model
			}
			if reasoning != "" {
				params["reasoning_effort"] = reasoning
			}
			switch strings.ToLower(fullReasoning) {
			case "":
				// untouched
			case "true":
				params["full_reasoning"] = true
			case "false":
				params["full_reasoning"] = false
			default:
				return fmt.Errorf("--full-reasoning must be 'true' or 'false'")
			}
			if len(params) == 0 {
				return fmt.Errorf("at least one of --provider, --model, --reasoning-effort, --full-reasoning is required")
			}
			return state.Client.SetToolLLMParams(cmd.Context(), args[0], params)
		},
	}
	cmd.Flags().StringVar(&provider, "provider", "", "LLM provider (e.g. anthropic, openai)")
	cmd.Flags().StringVar(&model, "model", "", "Model name")
	cmd.Flags().StringVar(&reasoning, "reasoning-effort", "", "low | medium | high")
	cmd.Flags().StringVar(&fullReasoning, "full-reasoning", "", "true | false")
	return cmd
}

func newToolsResetLLMCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "reset-llm <tool_id> <key> [key...]",
		Short: "Delete LLM-parameter overrides so code defaults apply",
		Args:  cobra.MinimumNArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.ResetToolLLMParams(cmd.Context(), args[0], args[1:])
		},
	}
}

func stringField(m map[string]any, k string) string {
	if v, ok := m[k]; ok {
		if s, ok := v.(string); ok {
			return s
		}
		return fmt.Sprintf("%v", v)
	}
	return ""
}

func boolField(m map[string]any, k string, fallback bool) string {
	if v, ok := m[k]; ok {
		if b, ok := v.(bool); ok {
			if b {
				return "yes"
			}
			return "no"
		}
	}
	if fallback {
		return "yes"
	}
	return "no"
}

func dumpJSON(v any) {
	b, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		fmt.Println(v)
		return
	}
	fmt.Println(string(b))
}

File: app/cli/commands/tools.py
"""`openpa tools ...` — list and configure tools & skills.

Mirrors `cli/cmd/tools.go`.
"""

from __future__ import annotations

import json as _json
import sys
from typing import Any, Optional

import typer

from app.cli.commands._helpers import graceful_errors


tools_app = typer.Typer(
    name="tools",
    help="List and configure tools & skills.",
    no_args_is_help=True,
)


@tools_app.command("list")
@graceful_errors
def tools_list(
    ctx: typer.Context,
    type_filter: Optional[str] = typer.Option(
        None,
        "--type",
        help="Filter by tool_type: built-in, mcp, a2a, skill, intrinsic.",
    ),
) -> None:
    """List all tools (built-in, mcp, a2a, skill, intrinsic)."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import list_tools
    from app.cli.config import Config
    from app.cli.output import OutputMode, Table, print_json
    from app.cli.output.formatting import bool_field, string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> list[dict[str, Any]]:
        async with Client(cfg) as client:
            return await list_tools(client, type_filter or "")

    tools = asyncio.run(_run())

    if mode.json:
        print_json(tools)
        return

    table = Table(mode, "TOOL_ID", "TYPE", "ENABLED", "CONFIGURED", "NAME")
    for row in tools:
        table.add_row(
            string_field(row, "tool_id"),
            string_field(row, "tool_type"),
            bool_field(row, "enabled", True),
            bool_field(row, "configured", False),
            string_field(row, "name"),
        )
    table.render()


@tools_app.command("get")
@graceful_errors
def tools_get(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Show detailed configuration for a tool."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import get_tool
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json, print_kv
    from app.cli.output.formatting import bool_field, string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await get_tool(client, tool_id)

    tool = asyncio.run(_run())

    if mode.json:
        print_json(tool)
        return

    print_kv([
        ("tool_id", string_field(tool, "tool_id")),
        ("name", string_field(tool, "name")),
        ("tool_type", string_field(tool, "tool_type")),
        ("description", string_field(tool, "description")),
        ("configured", bool_field(tool, "configured", False)),
    ])
    if isinstance(tool.get("config"), dict):
        sys.stdout.write("\n--- config ---\n")
        sys.stdout.write(_json.dumps(tool["config"], indent=2, ensure_ascii=False, default=str) + "\n")
    locked = tool.get("locked_llm_fields")
    if isinstance(locked, list) and locked:
        sys.stdout.write("\nlocked_llm_fields: " + ", ".join(str(v) for v in locked) + "\n")


@tools_app.command("enable")
@graceful_errors
def tools_enable(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Enable an A2A or MCP tool."""
    _set_tool_enabled(ctx, tool_id, True)


@tools_app.command("disable")
@graceful_errors
def tools_disable(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
) -> None:
    """Disable an A2A or MCP tool."""
    _set_tool_enabled(ctx, tool_id, False)


def _set_tool_enabled(ctx: typer.Context, tool_id: str, enabled: bool) -> None:
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import set_tool_enabled
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await set_tool_enabled(client, tool_id, enabled)

    asyncio.run(_run())


@tools_app.command("set-var")
@graceful_errors
def tools_set_var(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    pairs: list[str] = typer.Argument(
        ...,
        help="One or more KEY=VALUE pairs.",
        metavar="KEY=VALUE",
    ),
) -> None:
    """Set Tool Variables (env-style key/value pairs)."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import set_tool_variables
    from app.cli.config import Config

    if not pairs:
        typer.echo("expected at least one KEY=VALUE pair", err=True)
        raise typer.Exit(code=1)

    variables: dict[str, str] = {}
    for kv in pairs:
        if "=" not in kv:
            typer.echo(f"expected KEY=VALUE, got '{kv}'", err=True)
            raise typer.Exit(code=1)
        k, v = kv.split("=", 1)
        variables[k] = v

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await set_tool_variables(client, tool_id, variables)

    asyncio.run(_run())


@tools_app.command("set-args")
@graceful_errors
def tools_set_args(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    json_body: str = typer.Option(
        ...,
        "--json",
        help="Tool arguments as a JSON object.",
    ),
) -> None:
    """Set Tool Arguments from a JSON object."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import set_tool_arguments
    from app.cli.config import Config

    try:
        values = _json.loads(json_body)
    except _json.JSONDecodeError as e:
        typer.echo(f"invalid --json: {e}", err=True)
        raise typer.Exit(code=1) from e
    if not isinstance(values, dict):
        typer.echo("--json must be an object", err=True)
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await set_tool_arguments(client, tool_id, values)

    asyncio.run(_run())


@tools_app.command("set-llm")
@graceful_errors
def tools_set_llm(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    provider: Optional[str] = typer.Option(None, "--provider", help="LLM provider."),
    model: Optional[str] = typer.Option(None, "--model", help="Model name."),
    reasoning_effort: Optional[str] = typer.Option(
        None, "--reasoning-effort", help="low | medium | high."
    ),
    full_reasoning: Optional[str] = typer.Option(
        None, "--full-reasoning", help="true | false."
    ),
) -> None:
    """Set LLM Parameters for a tool (partial - omitted flags unchanged)."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import set_tool_llm_params
    from app.cli.config import Config

    params: dict[str, Any] = {}
    if provider:
        params["llm_provider"] = provider
    if model:
        params["llm_model"] = model
    if reasoning_effort:
        params["reasoning_effort"] = reasoning_effort
    if full_reasoning is not None and full_reasoning != "":
        v = full_reasoning.lower()
        if v == "true":
            params["full_reasoning"] = True
        elif v == "false":
            params["full_reasoning"] = False
        else:
            typer.echo("--full-reasoning must be 'true' or 'false'", err=True)
            raise typer.Exit(code=1)

    if not params:
        typer.echo(
            "at least one of --provider, --model, --reasoning-effort, --full-reasoning is required",
            err=True,
        )
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await set_tool_llm_params(client, tool_id, params)

    asyncio.run(_run())


@tools_app.command("reset-llm")
@graceful_errors
def tools_reset_llm(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    keys: list[str] = typer.Argument(
        ...,
        help="Keys to revert to code defaults.",
        metavar="KEY",
    ),
) -> None:
    """Delete LLM-parameter overrides so code defaults apply."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import reset_tool_llm_params
    from app.cli.config import Config

    if not keys:
        typer.echo("expected at least one key to reset", err=True)
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await reset_tool_llm_params(client, tool_id, keys)

    asyncio.run(_run())


@tools_app.command("register-long-running")
@graceful_errors
def tools_register_long_running(
    ctx: typer.Context,
    tool_id: str = typer.Argument(..., help="Tool id."),
    force: bool = typer.Option(
        False,
        "--force",
        help="Bypass the duplicate-command check.",
    ),
) -> None:
    """Spawn a skill's long_running_app and persist it as autostart."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.tools import register_long_running_app
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json, print_kv
    from app.cli.output.formatting import string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await register_long_running_app(client, tool_id, force)

    out = asyncio.run(_run())

    if mode.json:
        print_json(out)
        return
    print_kv([
        ("process_id", string_field(out, "process_id")),
        ("autostart_id", str(out.get("autostart_id", ""))),
        ("command", string_field(out, "command")),
        ("working_dir", string_field(out, "working_dir")),
    ])
